snapshot/btrfs/btrfs.go

```diff
@@ -35,37 +60,199 @@ func (b *Btrfs) Prepare(key, parent string) ([]containerd.Mount, error) {
 		}
 	} else {
 		// btrfs subvolume snapshot /parent /subvol
-		if err := btrfs.SubvolSnapshot(dir, parent, false); err != nil {
+		if err := btrfs.SubvolSnapshot(dir, parentp, readonly); err != nil {
+			return nil, err
+		}
+
+		if err := os.Symlink(parentp, parentlink); err != nil {
 			return nil, err
 		}
 	}
 
+	if err := ioutil.WriteFile(namep, []byte(key), 0644); err != nil {
+		return nil, err
+	}
+
+	return b.mounts(dir)
+}
+
+func (b *Driver) mounts(dir string) ([]containerd.Mount, error) {
+	var options []string
+
 	// get the subvolume id back out for the mount
 	info, err := btrfs.SubvolInfo(dir)
 	if err != nil {
 		return nil, err
 	}
 
+	options = append(options, fmt.Sprintf("subvolid=%d", info.ID))
+
+	if info.Readonly {
+		options = append(options, "ro")
+	}
+
 	return []containerd.Mount{
 		{
 			Type:   "btrfs",
 			Source: b.device, // device?
 			// NOTE(stevvooe): While it would be nice to use to uuids for
 			// mounts, they don't work reliably if the uuids are missing.
-			Options: []string{fmt.Sprintf("subvolid=%d", info.ID)},
+			Options: options,
 		},
 	}, nil
 }
 
-func (b *Btrfs) Commit(name, key string) error {
-	dir := filepath.Join(b.root, "active", hash(key))
+func (b *Driver) Commit(name, key string) error {
+	var (
```

Comment:
Agreed, if we go with this approach. I'm looking at using a common metadata store, as all of these links cost quite a few syscalls and make a mess of the filesystem. Technically, this driver now needs locking to be idempotent and concurrency safe.